bdio-protobuf/src/main/java/com/blackducksoftware/bdio/proto/ProtobufBdioReader.java

```diff
@@ -0,0 +1,204 @@
+				throw new RuntimeException("Unsupported header message type");
+			}
+		}
+
+		byte[] formatVersionBytes = new byte[2];
+		in.read(formatVersionBytes);
+		short formatVersion = Shorts.fromByteArray(formatVersionBytes);
+
+		if (formatVersion == 1) {
+			return ProtoScanHeader.parseFrom(in);
+		} else if (formatVersion == 2) {
+			Any any = Any.parseFrom(in);
+			if (any.is(ProtoScanHeader.class)) {
+				return any.unpack(ProtoScanHeader.class);
+			} else {
+				throw new RuntimeException("Unknown type for header data: " + any.getDescriptorForType().getFullName());
+			}
+		} else {
+			throw new RuntimeException("Unsupported header message version: " + formatVersion);
+		}
+	}
+
+	/**
+	 * Reads just the header document from the supplied BDIO ZIP archive.
+	 *
+	 * @param in the BDIO ZIP archive
+	 * @return the scan header
+	 * @throws IOException if the header could not be read from the supplied stream
+	 */
+	public static ProtoScanHeader readHeaderFromScanFile(ZipInputStream in) throws IOException {
+		byte[] bdioEntryTypeBytes = new byte[2];
+
+		while (in.getNextEntry() != null) {
+			in.read(bdioEntryTypeBytes);
+			short bdioEntryType = Shorts.fromByteArray(bdioEntryTypeBytes);
+
+			if (bdioEntryType == BdioEntryType.HEADER.ordinal()) {
+				return readHeaderChunk(in, false);
+			}
+		}
+
+		throw new RuntimeException("Header file not present in BDIO archive");
+	}
+
+	/**
+	 * Read bdio nodes from supplied input stream, representing bdio chunk (entry).
+	 *
+	 * @param in input stream
+	 * @return BdioChunk data structure containing deserialized nodes
+	 * @throws IOException
+	 */
+	public static BdioChunk readBdioChunk(InputStream in, boolean verifyType) throws IOException {
+		Set<ProtoFileNode> fileData = new HashSet<>();
+		Set<ProtoDependencyNode> dependencyData = new HashSet<>();
+		Set<ProtoComponentNode> componentData = new HashSet<>();
+		Set<ProtoAnnotationNode> annotationData = new HashSet<>();
+		Set<ProtoImageNode> imageData = new HashSet<>();
+		Set<ProtoImageLayerNode> layerData = new HashSet<>();
+
+		if (verifyType) {
+			byte[] bdioEntryTypeBytes = new byte[2];
+			in.read(bdioEntryTypeBytes);
+			short bdioEntryType = Shorts.fromByteArray(bdioEntryTypeBytes);
+
+			if (bdioEntryType != BdioEntryType.CHUNK.ordinal()) {
+				throw new RuntimeException("Unsupported bdio entry type");
+			}
+		}
+
+		readBdioNodes(in, dependencyData, componentData, fileData, annotationData, imageData, layerData);
+
+		return new BdioChunk(fileData, dependencyData, componentData, annotationData, imageData, layerData);
+	}
+
+	private static void readBdioNodes(InputStream in, Set<ProtoDependencyNode> dependencyData,
+			Set<ProtoComponentNode> componentData, Set<ProtoFileNode> fileData, Set<ProtoAnnotationNode> annotationData,
+			Set<ProtoImageNode> imageData, Set<ProtoImageLayerNode> layerData) throws IOException {
+
+		short version = readVersion(in);
+
+		if (version == 1) {
```

Comment:
It seems likely that we will inevitably have a version 3, version 4, etc.

This is inevitably going to spaghetti the code over time.   Having a class that provides a versioned ProtobufBdioReader implementation based on the identified version could help isolate these code differences and make it simpler to test and extend over time.

Reply:
Refactored the code to have separate reader per version that uses common functionality from the base class.